src/main/scala/uk/sky/fs2/kafka/topicloader/TopicLoader.scala

```diff
@@ -34,56 +34,67 @@ trait TopicLoader {
+    * by the requested strategy.
+    *
+    * @param topics
+    *   topics to load
+    * @param strategy
+    *   A [[LoadTopicStrategy]]
+    * @param consumerSettings
+    *   [[fs2.kafka.ConsumerSettings]] for the given topics
+    */
   def load[F[_] : Async : LoggerFactory, K, V](
       topics: NonEmptyList[String],
       strategy: LoadTopicStrategy,
       consumerSettings: ConsumerSettings[F, K, V]
   ): Stream[F, ConsumerRecord[K, V]] =
     KafkaConsumer
       .stream(consumerSettings)
-      .evalMap { consumer =>
-        {
-          for {
-            logOffsets <- OptionT(logOffsetsForTopics(topics, strategy, consumer))
-            _          <- OptionT.liftF(
-                            consumer.assign(logOffsets.keys) *>
-                              logOffsets.toNel.traverse { case (tp, o) => consumer.seek(tp, o.lowest) }
-                          )
-          } yield load(consumer, logOffsets)
-        }.getOrElse(Stream.empty)
-      }
-      .flatten
-
+      .flatMap(load(topics, strategy, _))
+
+  /** Stream that loads the specified topics from the beginning. When the latest current offsets are reached, the
+    * `onLoad` callback is evaluated, and the stream continues.
+    *
+    * @param topics
+    *   topics to load
+    * @param consumerSettings
+    *   [[fs2.kafka.ConsumerSettings]] for the given topics
+    * @param onLoad
+    *   A callback that will be evaluated on once the current offsets are reached
+    */
   def loadAndRun[F[_] : Async : LoggerFactory, K, V](
       topics: NonEmptyList[String],
       consumerSettings: ConsumerSettings[F, K, V]
   )(onLoad: Resource.ExitCase => F[Unit]): Stream[F, ConsumerRecord[K, V]] =
     KafkaConsumer
       .stream(consumerSettings)
-      .evalMap { consumer =>
-        {
-          for {
-            logOffsets    <- OptionT(logOffsetsForTopics(topics, LoadAll, consumer))
-            _             <- OptionT.liftF(
-                               consumer.assign(logOffsets.keys) *>
-                                 logOffsets.toNel.traverse { case (tp, o) => consumer.seek(tp, o.lowest) }
-                             )
-            preLoadStream <- OptionT.pure(load(consumer, logOffsets))
-          } yield preLoadStream.onFinalizeCase(onLoad) ++ consumer.records.map(_.record)
-        }.getOrElse(Stream.empty)
+      .flatMap { consumer =>
+        load(topics, LoadAll, consumer).onFinalizeCase(onLoad) ++ consumer.records.map(_.record)
       }
-      .flatten
 
   private def load[F[_] : Async : LoggerFactory, K, V](
-      consumer: KafkaConsumer[F, K, V],
-      logOffsets: NonEmptyMap[TopicPartition, LogOffsets]
-  ): Stream[F, ConsumerRecord[K, V]] = consumer.records.map(_.record).through(filterBelowHighestOffset(logOffsets))
+      topics: NonEmptyList[String],
+      strategy: LoadTopicStrategy,
+      consumer: KafkaConsumer[F, K, V]
+  ): Stream[F, ConsumerRecord[K, V]] = Stream
+    .eval({
+      for {
```

Comment:
why has all this moved from `loadAndRun` to here? i can read the code of course but i just don't really understand the changes

Reply:
Both load and loadAndRun were doing the same thing, but loadAndRun was hooking onto the finalize and appending a new stream. Thought it made sense to pull out the load stream so both load and loadAndRun can use it

Reply:
what you say makes sense, its just surprising me that they do the same thing but the previous impl of `load` was so simple

Reply:
I think the previous impl was slightly wrong, and more has to be done in the `load` method now (like assign all partitions and seek to beginning)